pkg/plugins/docker/dockerfile/mobyparser/main.go

```diff
@@ -0,0 +1,199 @@
+	}
+
+	found, val, err := m.replace(data.AST)
+	if err != nil {
+		logrus.Error(err.Error())
+		return false
+	}
+
+	if !found {
+		logrus.Infof("\u2717 Instruction %q wasn't found", m.Instruction)
+		return false
+	}
+
+	if val == m.Value {
+		logrus.Infof("\u2714 Instruction %q is correctly set to %q",
+			m.Instruction,
+			m.Value)
+		return true
+	}
+
+	logrus.Infof("\u2717 Instruction %q found but incorrectly set to %q instead of %q",
+		m.Instruction,
+		val,
+		m.Value)
+	return true
+}
+
+func (m MobyParser) ReplaceInstructions(dockerfileContent []byte, sourceValue string) ([]byte, types.ChangedLines, error) {
+
+	changed := make(types.ChangedLines)
+
+	data, err := parser.Parse(bytes.NewReader(dockerfileContent))
+	if err != nil {
+		return dockerfileContent, changed, err
+	}
+
+	m.Value = sourceValue
+
+	valueFound, oldVersion, err := m.replace(data.AST)
+	if err != nil {
+		return dockerfileContent, changed, err
+	}
+
+	if valueFound {
+		if oldVersion == m.Value {
+			logrus.Infof("\u2714 Instruction %q already set to %q, nothing else need to be done",
+				m.Instruction,
+				m.Value)
+			return dockerfileContent, changed, nil
+		}
+
+		logrus.Infof("\u2714 Instruction %q, was updated from %q to %q",
+			m.Instruction,
+			oldVersion,
+			m.Value,
+		)
+	} else {
+		return dockerfileContent, changed, fmt.Errorf("\u2717 cannot find instruction %q", m.Instruction)
+	}
+
+	newDockerfileContent := ""
+	err = Marshal(data, &newDockerfileContent)
+	if err != nil {
+		return dockerfileContent, changed, err
+	}
+
+	originalDockerfileContentLines := strings.Split(string(dockerfileContent), "\n")
+	newDockerfileContentLines := strings.Split(newDockerfileContent, "\n")
+	for index, currentLine := range newDockerfileContentLines {
+		// If the line is not a comment, then we expect it to be a change
+		if !strings.HasPrefix(strings.TrimSpace(currentLine), "#") {
+			// Check that the originalDockerfile has the original line or ignore the change
+			if len(originalDockerfileContentLines) > index && currentLine != originalDockerfileContentLines[index] {
+				changed[index] = types.LineDiff{Original: originalDockerfileContentLines[index], New: currentLine}
+			}
+		}
+	}
+
+	return []byte(newDockerfileContent), changed, nil
+}
+
+func (m MobyParser) String() string {
+	return fmt.Sprintf("%s %s", m.Instruction, m.Value)
+}
+
+// isPositionKey checks if key use the array position like instruction[0][0].
+// where first [0] references the instruction position
+// and the  second [0] references the element inside instruction line
+func isPositionKeys(key string) bool {
+	matched, err := regexp.MatchString(`^(.*)\[[[:digit:]]*\]\[[[:digit:]]*\]$`, key)
+
+	if err != nil {
+		logrus.Errorf("err - %s", err)
+	}
+	return matched
+}
+
+func getPositionKeys(k string) (
+	key string,
+	instructionPosition int,
+	elementPosition int,
+	err error) {
+
+	if isPositionKeys(k) {
+		re := regexp.MustCompile(`(.*)*\[[[:digit:]]*\]\[[[:digit:]]*\]$`)
```

Comment:
Compiling a regexp takes time and you might want to avoid doing this each time you want it as an instance of a `Regexp` can be reused. 

I would suggest do store the compiled regexps in a private package level variable.

For example: 

```
var positionRegex = regexp.MustCompile(`(.*)*\[[[:digit:]]*\]\[[[:digit:]]*\]$`)

func getPositionKeys(k string) {
        // ...
        keys := positionRegex.FindStringSubmatch(k)
}
```

Does it makes any sense ?

Reply:
@jlevesy it definitely makes sense thanks for the suggestion

Reply:
Thanks for the suggestion I just moved them at the package level

Reply:
^ @dduportal